File: algorithms/2_medium/L/01493_longest_subarray_of_1s_after_deleting_one_element/sliding_window.py
class Solution:
    def longestSubarray(self, nums: List[int]) -> int:

        left = 0
        zero_count = 0
        max_subarr_len = 0
        nums_len = len(nums)

        for right in range(nums_len):
            if nums[right] == 0:
                zero_count += 1

            while zero_count > 1:
                if nums[left] == 0:
                    zero_count -= 1

                left += 1

            max_subarr_len = max(max_subarr_len, right - left + 1 - zero_count)

        # Edge case: if entire arr is 1, then max_subarr_len will be equal to len
        # need to delete at least one of them
        if max_subarr_len == nums_len:
            return max_subarr_len - 1

        return max_subarr_len

File: algorithms/2_medium/L/01493_longest_subarray_of_1s_after_deleting_one_element/sliding_window.rs
impl Solution {
    pub fn longest_subarray(nums: Vec<i32>) -> i32 {

        let mut left = 0_usize;
        let mut zero_count = 0_usize;
        let mut max_subarr_len = 0_usize;
        let len = nums.len();

        for right in 0..len {
            if nums[right] == 0 {
                zero_count += 1;
            }

            while zero_count > 1 {
                if nums[left] == 0 {
                    zero_count -= 1;
                }

                left += 1;
            }

            max_subarr_len = max_subarr_len.max(right - left + 1 - zero_count);
        }

        // Edge case: if entire arr is 1, then max_subarr_len will be equal to len
        // need to delete at least one of them
        if max_subarr_len == len {
            return max_subarr_len as i32 - 1;
        }

        max_subarr_len as i32
    }
}
